### PR TITLE
Ensure safeTransferFrom reverts with expected message

### PR DESCRIPTION
**Pull request by wilsoncusack**
Currently, if the receiving contract does not implement `onERC721Received`, the tx will fail without the "UNSAFE_RECIPIENT". This PR make the `onERC721Received` in a try catch to ensure we surface our revert message as intended.

### CODE REVIEW

**Comment by transmissions11** [comment]
mind inlining this? hehe

https://twitter.com/transmissions11/status/1477775750145138691?s=20

**Comment by wilsoncusack** [comment]
> mind inlining this? hehe
> 
> https://twitter.com/transmissions11/status/1477775750145138691?s=20

Done. From forge gas it looks like the failed case is now ~300 gas more expensive, but the happy case is slightly cheaper than before! Would be curious to see a dapp tools snapshot.

File: src/tokens/ERC721.sol [hunk 1]
```diff
@@ -102,7 +102,19 @@ abstract contract ERC721 {
...
+                require(retval == ERC721TokenReceiver.onERC721Received.selector, "UNSAFE_RECIPIENT");
+            } catch (bytes memory reason) {
+                if (reason.length == 0) {
+                    revert("UNSAFE_RECIPIENT");
+                } else {
+                    assembly {
+                        revert(add(32, reason), mload(reason))
+                    }
+                }
+            }
```
Comment (transmissions11) [inline src/tokens/ERC721.sol:116]: lets just always revert with UNSAFE_RECIPIENT, idgaf about bubbling up their message

**Comment by transmissions11** [comment]
perfect!

File: src/tokens/ERC721.sol [hunk 2]
```diff
@@ -102,12 +102,13 @@ abstract contract ERC721 {
     ) public virtual {
         transferFrom(from, to, id);
 
-        require(
-            to.code.length == 0 ||
-                ERC721TokenReceiver(to).onERC721Received(msg.sender, from, id, "") ==
-                ERC721TokenReceiver.onERC721Received.selector,
-            "UNSAFE_RECIPIENT"
-        );
+        if(to.code.length != 0){
```
Comment (lastperson) [inline src/tokens/ERC721.sol:105]:
```suggestion
        if (to.code.length != 0) {
```

File: src/tokens/ERC721.sol [hunk 3]
```diff
@@ -118,12 +119,13 @@ abstract contract ERC721 {
     ) public virtual {
         transferFrom(from, to, id);
 
-        require(
-            to.code.length == 0 ||
-                ERC721TokenReceiver(to).onERC721Received(msg.sender, from, id, data) ==
-                ERC721TokenReceiver.onERC721Received.selector,
-            "UNSAFE_RECIPIENT"
-        );
+        if(to.code.length != 0){
```
Comment (lastperson) [inline src/tokens/ERC721.sol:122]:
```suggestion
        if (to.code.length != 0) {
```

**Review by lastperson — COMMENTED** [review]
Just passing by.

**Comment by wilsoncusack** [comment]
Ser @transmissions11 do you want to merge or should I close

**Comment by transmissions11** [comment]
hmmm after more thinking we already use unintuitive revert messages for plenty of things, so will keep as is on master. tho i cleaned up this PR in case anyone wants to use it :)

**Pull request closed without merging**